### DISCUSSION TITLE
How to add interceptor to useFetch() ?

### DISCUSSION THREAD

**Discussion by derarsat** [nuxt/nuxt#15913] [Questions] [answered]
Hello
I need a solution to globally add interceptor to useFetch() composable.
any ideas ?

**Comment by danielroe** [comment]
You can create your own `useFetch`:

```js
// ~/composables/useMyFetch.ts
export const useMyFetch = (id: string) => {
  return useFetch(id, { /* your options */ })
}
```

**Reply by Diizzayy** [reply]
@danielroe I hadn't realized you beat me to it 😂

**Comment by Diizzayy** [comment, marked as answer]
@derar-toothpick You can create a custom fetch composable and utilize the [ohmyfetch interceptors ](https://github.com/unjs/ohmyfetch#%EF%B8%8F-interceptors)

`nuxt-app/composables/useCustomFetch.ts`
```ts
export const useCustomFetch = (url: string, options?: FetchOptions) => {
  return useFetch(url, {
    ...options,
    async onResponse({ request, response, options }) {
      console.log('[fetch response]')
    },
    async onResponseError({ request, response, options }) {
      console.log('[fetch response error]')
    },

    async onRequest({ request, options }) {
      console.log('[fetch request]')
    },
    async onRequestError({ request, options, error }) {
      console.log('[fetch request error]')
    }
  })
}
```

**Reply by derarsat** [reply]
Working !!

**Reply by PavlosMac** [reply]
Do you use typescript, I find it difficult to wrap useFetch and still get the return types, data refresh pending and such?

**Reply by BenWaNH** [reply]
Same issue than @PavlosMac 
I try to wrap useFetch like this. But when i pick props in options types is not return correctly...

**Reply by skpstg** [reply]
Did anyone find solution of type issues with composable wrapper for useFetch?

**Reply by flihub** [reply]
```javascript
import { UseFetchOptions } from "nuxt/dist/app/composables";

export const useFetchApi = (url: string, options?: UseFetchOptions<object>) => {
    return useFetch(url, {
        ...options,
        async onRequest({ request, options }) {
            console.log("[fetch request]");

            const headers = new Headers(options.headers);
            headers.set("Authorization", "Bearer Token");

            options.headers = headers;
        },
        async onRequestError({ request, options, error }) {
            console.log("[fetch request error]");
        },
        async onResponse({ request, response, options }) {
            console.log("[fetch response]");
        },
        async onResponseError({ request, response, options }) {
            console.log("[fetch response error]");
        },
    });
};

```

**Reply by jimoh-sodiq** [reply]
Any solution with full type handling for return types too?

**Reply by christianholland** [reply]
@Diizzayy 
Given that the `options` object already contains the `onResponse` interceptor it will get overwritten by the other `onResponse` definition. Is there a way to extent the `onResponse` logic instead of overwriting?

**Reply by marcaureln** [reply]
For those having type issues (like @jimoh-sodiq and myself), this answer handle it very well: https://github.com/nuxt/nuxt/discussions/16715#discussioncomment-6564177

I had some issue with this import `import type { KeysOf, PickFrom } from 'nuxt/dist/app/composables/asyncData'`
But, I just copied the types from node_module:
```ts
export type PickFrom<T, K extends Array<string>> = T extends Array<any> ? T : T extends Record<string, any> ? keyof T extends K[number] ? T : K[number] extends never ? T : Pick<T, K[number]> : T;
export type KeysOf<T> = Array<T extends T ? (keyof T extends string ? keyof T : never) : never>;
```

**Reply by harryisfish** [reply]
A simple one: [use-custom-fetch-composable](https://nuxt.com/docs/examples/advanced/use-custom-fetch-composable)

**Comment by MatthD** [comment]
Hello, I am trying to check how to work with interceptor on nuxt-bridge , is it possible ?

**Comment by Matman94** [comment]
I believe that a solution such as 'own composable' that is actually a decorator is not a solution....
I would like something global per request and response like it was in nuxt2.
What if someone on the team forgets/doesn't know about using a specially created composable?

**Reply by cannap** [reply]
maybe tell them before they use it or nuxt should make a option to rename the useFetch/and others to something

**Comment by shba007** [comment]
How can we use $fetch.create to create a composable as interceptor also use useRuntimeConfig() inside it

**Reply by shba007** [reply]
Solved [#146](https://github.com/unjs/ofetch/issues/146#issuecomment-1322303318)

**Comment by cannap** [comment]
whats about an option to rename the "fetch" composables so other ppl can call them "useFetch" but in the background its the customized one

**Comment by Daisigu** [comment]
Maybe it will be useful to someone
https://www.npmjs.com/package/nuxt-advanced-fetch?activeTab=readme
